File: mac/Marquee/Features/Settings/LastActiveLabel.swift
import Foundation

// "Active 3 hours ago" under each member in Settings → Household members.
// A port of the website's lib/users/last-active-label.ts: same wording, same
// thresholds. Anything within the last ten minutes is "Active now" (the
// server records activity to within 5 minutes).

/// Pure; unit tested. `nil` means the account has never been used.
func lastActiveLabel(
    _ lastActiveAt: Date?,
    now: Date,
    timeZone: TimeZone = .current
) -> String {
    guard let lastActiveAt else { return "Never signed in" }
    let minute: TimeInterval = 60
    let hour = 60 * minute
    let day = 24 * hour
    let ago = now.timeIntervalSince(lastActiveAt)
    if ago < 10 * minute { return "Active now" }
    if ago < hour { return "Active \(Int((ago / minute).rounded(.down))) minutes ago" }
    if ago < day {
        let hours = Int((ago / hour).rounded(.down))
        return "Active \(hours) \(hours == 1 ? "hour" : "hours") ago"
    }
    if ago < 2 * day { return "Active yesterday" }
    if ago < 30 * day { return "Active \(Int((ago / day).rounded(.down))) days ago" }
    let formatter = DateFormatter()
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.timeZone = timeZone
    formatter.dateFormat = "MMM d, yyyy"
    return "Last active \(formatter.string(from: lastActiveAt))"
}

File: lib/users/last-active-label.ts
// "Active 3 hours ago" under each member in Settings → Household members.
// Kept to the recording interval's precision (lib/users/last-active.ts):
// anything within the last ten minutes is "Active now".

const MINUTE = 60 * 1000;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;

/** Pure; unit tested. `null` means the account has never been used. */
export function lastActiveLabel(lastActiveAt: Date | null, now: Date): string {
  if (!lastActiveAt) return "Never signed in";
  const ago = now.getTime() - lastActiveAt.getTime();
  if (ago < 10 * MINUTE) return "Active now";
  if (ago < HOUR) return `Active ${Math.floor(ago / MINUTE)} minutes ago`;
  if (ago < DAY) {
    const hours = Math.floor(ago / HOUR);
    return `Active ${hours} ${hours === 1 ? "hour" : "hours"} ago`;
  }
  if (ago < 2 * DAY) return "Active yesterday";
  if (ago < 30 * DAY) return `Active ${Math.floor(ago / DAY)} days ago`;
  const date = lastActiveAt.toLocaleDateString("en-US", { month: "short", day: "numeric", year: "numeric" });
  return `Last active ${date}`;
}
